Sources/SourceKitLSP/Swift/ReferenceDocument.swift

```diff
@@ -0,0 +1,150 @@
+//===----------------------------------------------------------------------===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2024 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See https://swift.org/LICENSE.txt for license information
+// See https://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+
+import Foundation
+import LanguageServerProtocol
+
+/// A Reference Document is a document whose url scheme is `sourcekit-lsp:` and whose content can only be retrieved
+/// using `GetReferenceDocumentRequest`.
+///
+/// The `url` will be of the form: `sourcekit-lsp://<documentType>/<displayName>?<data>`
+/// Here,
+///  - The `documentType` denotes the kind of the content present in the reference document
+///  - The `data` denotes the parameter-value pairs such as "p1=v1&p2=v2&..." needed to generate
+/// the content of the reference document.
+///  - The `displayName` is the displayed file name of the reference document. It doesn't involve in generating
+/// the content of the reference document.
+package protocol ReferenceDocument {
+  /// A type to represent the parameter-value pairs of `data` as an object
+  associatedtype Data: ReferenceDocumentData
+
+  var url: URL { get throws }
+
+  static var documentType: ReferenceDocumentType { get }
+
+  var displayName: String { get }
+  var data: Data { get set }
+
+  init(from url: URL) throws
+  init(from urlString: String) throws
+
+  init(data: Data)
+
+  /// Creates the `DocumentURI` of the source file, using which the reference document is generated, and returns it.
+  /// - Returns: The `DocumentURI` of the source file
+  func sourceDocument() -> DocumentURI
+
+  /// This generates the content of the reference document and returns it.
+  /// - Parameter languageService: The `SwiftLanguageService` using which it generates
+  /// the content of the reference document.
+  /// - Returns: The content of the reference document
+  func documentContent(using languageService: SwiftLanguageService) async throws -> String
```

Comment:
I don’t like how `ReferenceDocument` is used to both form/parse the URL of a reference document and to get the document contents because it’s not clear what the type’s purpose is. I would prefer if we had a dedicated type to just do the URL conversions and then a method on `SwiftLanguageService` that is responsible for generating the document contents.

And since we know all the types of reference documents that SourceKit-LSP can support, it would probably be cleaner if we have an enum that represents all of them. For example, something like

```swift
enum ReferenceDocumentURL {
  case macroExpansion(MacroExpansionReferenceDocumentURLData)

  var url: URL { … }

  init(from url: URL) throws { … }

  // I don’t think any other members are needed
}
```